[SPARK-32417] Fix flakyness of BlockManagerDecommissionIntegrationSuite

agrawaldevesh opened the pull request:
### What changes were proposed in this pull request?

This test tries to fix the flakyness of BlockManagerDecommissionIntegrationSuite. 

### Description of the problem

Make the block manager decommissioning test be less flaky

An interesting failure happens when migrateDuring = true (and persist or shuffle is true):
- We schedule the job with tasks on executors 0, 1, 2.
- We wait 300 ms and decommission executor 0.
- If the task is not yet done on executor 0, it will now fail because
   the block manager won't be able to save the block. This condition is
   easy to trigger on a loaded machine where the github checks run.
- The task with retry on a different executor (1 or 2) and its shuffle
   blocks will land there.
- No actual block migration happens here because the decommissioned
   executor technically failed before it could even produce a block.

To remove the above race, this change replaces the fixed wait for 300 ms to wait for an actual task to succeed. When a task has succeeded, we know its blocks would have been written for sure and thus its executor would certainly be forced to migrate those blocks when it is decommissioned.

The change always decommissions an executor on which a real task finished successfully instead of picking the first executor. Because the system may choose to schedule nothing on the first executor and instead run the two tasks on one executor.

### Why are the changes needed?

I have had bad luck with BlockManagerDecommissionIntegrationSuite and it has failed several times on my PRs. So fixing it.

### Does this PR introduce _any_ user-facing change?

No, unit test only change.

### How was this patch tested?

Github checks. Ran this test 100 times, 10 at a time in parallel in a script.

dagrawal3409 committed:
Make the block manager decommissioning test be less flaky

An interesting failure happens when migrateDuring = true (and persist or
shuffle is true):

- We schedule the job with tasks on executors 0, 1, 2.
- We wait 300 ms and decommission executor 0.
- If the task is not yet done on executor 0, it will now fail because
  the block manager won't be able to save the block. This condition is
  easy to trigger on a loaded machine where the github checks run.
- The task with retry on a different executor (1 or 2) and its shuffle
  blocks will land there.
- No actual block migration happens here because the decommissioned
  executor technically failed before it could even produce a block.

So this change makes two fixes to remove the above race condition.
- When migrateDuring = true, wait for a task to complete and write the
  block, and then decommission that executor.
- When migrateDuring = false, it is still possible (because of delay
  scheduling) for two tasks to be run on the same executor serially and
  one executor to go idle. In which case, we must make sure to
  decommission an executor that actually had a task run on it.

SparkQA commented:
**[Test build #126517 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126517/testReport)** for PR 29226 at commit [`7659d22`](https://github.com/apache/spark/commit/7659d22bcc3e2a423a1068dd15d9b11eba76c2df).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31136/
Test PASSed.

agrawaldevesh commented:
cc: @holdenk @attilapiros for review please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31137/
Test PASSed.

SparkQA commented:
**[Test build #126518 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126518/testReport)** for PR 29226 at commit [`558702e`](https://github.com/apache/spark/commit/558702e6f55d6c17439d1942000bea08acde0b29).

holdenk commented:
Oh awesome, thanks for fixing this. I totally misdiagnosed this issue
(thought it was around placement of the partitions).

On Fri, Jul 24, 2020 at 5:36 PM Apache Spark QA <notifications@github.com>
wrote:

> *Test build #126518 has started
> <https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126518/testReport>*
> for PR 29226 at commit 558702e
> <https://github.com/apache/spark/commit/558702e6f55d6c17439d1942000bea08acde0b29>
> .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/spark/pull/29226#issuecomment-663784088>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAOT5OSXJNHXNXHEYMHJALR5ISHXANCNFSM4PHFIUWQ>
> .
>
-- 
Cell : 425-233-8271

SparkQA commented:
**[Test build #126517 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126517/testReport)** for PR 29226 at commit [`7659d22`](https://github.com/apache/spark/commit/7659d22bcc3e2a423a1068dd15d9b11eba76c2df).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126517/
Test PASSed.

SparkQA commented:
**[Test build #126518 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126518/testReport)** for PR 29226 at commit [`558702e`](https://github.com/apache/spark/commit/558702e6f55d6c17439d1942000bea08acde0b29).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126518/
Test PASSed.

agrawaldevesh commented:
Sorry it didn't work. I will make another go at it.

SparkQA commented:
**[Test build #126526 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126526/testReport)** for PR 29226 at commit [`288c34e`](https://github.com/apache/spark/commit/288c34ebaaec481f0fdf0db1938cf39e3970eb6d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31145/
Test PASSed.

SparkQA commented:
**[Test build #126526 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126526/testReport)** for PR 29226 at commit [`288c34e`](https://github.com/apache/spark/commit/288c34ebaaec481f0fdf0db1938cf39e3970eb6d).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126526/
Test FAILed.

SparkQA commented:
**[Test build #126530 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126530/testReport)** for PR 29226 at commit [`4c8bafb`](https://github.com/apache/spark/commit/4c8bafb9d55e54332a83d97fcbe8662a20116215).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31149/
Test PASSed.

SparkQA commented:
**[Test build #126531 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126531/testReport)** for PR 29226 at commit [`199ea03`](https://github.com/apache/spark/commit/199ea03f028804dc12039bbca4f25a7138a378a8).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31150/
Test PASSed.

SparkQA commented:
**[Test build #126536 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126536/testReport)** for PR 29226 at commit [`b2c2fad`](https://github.com/apache/spark/commit/b2c2fad1eb1caeeb5d6e135220398ad43a963858).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31155/
Test PASSed.

SparkQA commented:
**[Test build #126530 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126530/testReport)** for PR 29226 at commit [`4c8bafb`](https://github.com/apache/spark/commit/4c8bafb9d55e54332a83d97fcbe8662a20116215).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126530/
Test FAILed.

SparkQA commented:
**[Test build #126531 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126531/testReport)** for PR 29226 at commit [`199ea03`](https://github.com/apache/spark/commit/199ea03f028804dc12039bbca4f25a7138a378a8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126531/
Test PASSed.

SparkQA commented:
**[Test build #126536 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126536/testReport)** for PR 29226 at commit [`b2c2fad`](https://github.com/apache/spark/commit/b2c2fad1eb1caeeb5d6e135220398ad43a963858).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126536/
Test PASSed.

agrawaldevesh commented:
The remaining failure is in SparkR and is probably environment related.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31170/
Test PASSed.

SparkQA commented:
**[Test build #126550 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126550/testReport)** for PR 29226 at commit [`ba12869`](https://github.com/apache/spark/commit/ba12869cd1d87c54566a2584786763aff20e80bf).

SparkQA commented:
**[Test build #126550 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126550/testReport)** for PR 29226 at commit [`ba12869`](https://github.com/apache/spark/commit/ba12869cd1d87c54566a2584786763aff20e80bf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126550/
Test PASSed.

SparkQA commented:
**[Test build #126554 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126554/testReport)** for PR 29226 at commit [`a939c67`](https://github.com/apache/spark/commit/a939c675dd70ec0368e7d9d8f08f1ffca662cf1d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31174/
Test PASSed.

SparkQA commented:
**[Test build #126554 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126554/testReport)** for PR 29226 at commit [`a939c67`](https://github.com/apache/spark/commit/a939c675dd70ec0368e7d9d8f08f1ffca662cf1d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126554/
Test PASSed.

attilapiros commented on core/src/test/scala/org/apache/spark/storage/BlockManagerDecommissionIntegrationSuite.scala:
```diff
@@ -57,6 +58,11 @@ class BlockManagerDecommissionIntegrationSuite extends SparkFunSuite with LocalS
       .set(config.STORAGE_DECOMMISSION_ENABLED, true)
       .set(config.STORAGE_DECOMMISSION_RDD_BLOCKS_ENABLED, persist)
       .set(config.STORAGE_DECOMMISSION_SHUFFLE_BLOCKS_ENABLED, shuffle)
+      // Force exactly one executor per worker such that all block managers
+      // get the shuffle and RDD blocks.
+      .set(config.EXECUTOR_CORES.key, "1")
+      .set(config.CPUS_PER_TASK.key, "1")
+      .set(config.EXECUTOR_MEMORY.key, "1024m")
```
I think these changes are redundant and not needed as they are setting the same values as the defaults for these configs:

https://github.com/apache/spark/blob/a939c675dd70ec0368e7d9d8f08f1ffca662cf1d/core/src/main/scala/org/apache/spark/internal/config/package.scala#L293-L302

https://github.com/apache/spark/blob/a939c675dd70ec0368e7d9d8f08f1ffca662cf1d/core/src/main/scala/org/apache/spark/internal/config/package.scala#L524-L525

attilapiros commented on core/src/test/scala/org/apache/spark/storage/BlockManagerDecommissionIntegrationSuite.scala:
```diff
@@ -107,6 +115,21 @@ class BlockManagerDecommissionIntegrationSuite extends SparkFunSuite with LocalS
 
       override def onBlockUpdated(blockUpdated: SparkListenerBlockUpdated): Unit = {
+        if (blockUpdated.blockUpdatedInfo.blockId.isRDD && persist) {
+          // Persisted RDD blocks are a bit weirder than shuffle blocks: Even though
+          // the tasks are run say on executors (0, 1, 2), the RDD blocks might end up only
+          // on executors 0 and 1. So we cannot just indiscriminately decommission any executor.
```
Please investigate why this placement is happening (as at first it sounds a bit strange). 
You might find a more important bug to fix or very good lesson about the internals of RDD persistence.

Tipp: please double check "though the tasks are run say on executors (0, 1, 2)" part!

attilapiros reviewed (changes requested):
This PR might resolve the issue at hand but during the problem analysis it stops at a level and fails to reveal the root cause. I would like to encourage you to look a bit deeper: there might be an easier fix or a more straightforward one.

attilapiros commented on core/src/test/scala/org/apache/spark/storage/BlockManagerDecommissionIntegrationSuite.scala:
```diff
@@ -107,6 +115,21 @@ class BlockManagerDecommissionIntegrationSuite extends SparkFunSuite with LocalS
+          // executors and thus any executor can be decommissioned when `persist` is false.
+          val candidateExecToDecom = blockUpdated.blockUpdatedInfo.blockManagerId.executorId
+          if (execToDecommission.compareAndSet(null, candidateExecToDecom)) {
+            val decomContext = s"Decommissioning executor ${candidateExecToDecom} for persist"
+            logInfo(decomContext)
+            sched.decommissionExecutor(candidateExecToDecom,
+              ExecutorDecommissionInfo(decomContext, false))
```
duplicated code

SparkQA commented:
**[Test build #126582 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126582/testReport)** for PR 29226 at commit [`2cea5f4`](https://github.com/apache/spark/commit/2cea5f4673b746d068bf69f6818f23d1962039ca).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31205/
Test PASSed.

SparkQA commented:
**[Test build #126583 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126583/testReport)** for PR 29226 at commit [`913e6d7`](https://github.com/apache/spark/commit/913e6d72c0bb3b18739b730a0634bcbc35444437).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31206/
Test PASSed.

agrawaldevesh commented:
@attilapiros Thank you for pushing me to do the right thing here and go that extra mile ! I think I have now finally understood why this race condition was happening and I have fixed the PR description and the code accordingly. The fix hasn't changed fundamentally but I now know the right reasons behind it :-)

SparkQA commented:
**[Test build #126582 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126582/testReport)** for PR 29226 at commit [`2cea5f4`](https://github.com/apache/spark/commit/2cea5f4673b746d068bf69f6818f23d1962039ca).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126582/
Test PASSed.

SparkQA commented:
**[Test build #126583 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126583/testReport)** for PR 29226 at commit [`913e6d7`](https://github.com/apache/spark/commit/913e6d72c0bb3b18739b730a0634bcbc35444437).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126583/
Test FAILed.

attilapiros commented:
We are still not there yet.

So according to you analyses:
> - We wait 300 ms and decommission executor 0.
>
> - If the task is not yet done on executor 0, it will now fail because
the block manager won't be able to save the block. This condition is
easy to trigger on a loaded machine where the github checks run.

If we would need this block to be saved then there would be a much simpler fix by checking when the block appears in the `blocksUpdated` collection. But this goes against the `migrateDuring` original intention (see the first comment):

https://github.com/apache/spark/blob/913e6d72c0bb3b18739b730a0634bcbc35444437/core/src/test/scala/org/apache/spark/storage/BlockManagerDecommissionIntegrationSuite.scala#L157-L164

HyukjinKwon commented:
@agrawaldevesh, shell we file a JIRA? I think it's better to file a JIRA to track in general.

attilapiros commented:
@agrawaldevesh I gave some more thought: it must be really about the middle of the job so to wait for task finish is fine here.

I suggest to go for the much simpler fix by checking when the block appears in the `blocksUpdated` collection and trigger the decommissioning only after we have this block on the disk.

agrawaldevesh commented:
> If we would need this block to be saved then there would be a much simpler fix by checking when the block appears in the `blocksUpdated` collection. ~But this goes against the `migrateDuring` original intention (see the first comment):~

I considered this approach (among others) but decided against it. Please let me explain my reasoning:

- First, while I could always poll for the block to show up in blocksUpdated, you would agree that it is a lot of machinery particularly when we know that the blocksUpdated is placed by the `onBlockUpdated` listener call. So I decided to just trigger this decommissioning directly in the listener.

- Second question is whether the decommissioning be done in `onBlockUpdated` or `onTaskEnd`. I chose `onTaskEnd` because that is the point of no return for the block to be "live". Consider what happens when a result task fails after the block is updated but before it can be considered "done". The block it wrote would get garbage collected. This garbage collection would race with the decommissioning. It's possible that the migration starts after the GC and does not find the block to replicated. This would fail the assertion below.

agrawaldevesh commented:
> @agrawaldevesh, shell we file a JIRA? I think it's better to file a JIRA to track in general.

Turns out I was lucky: @gaborgsomogyi had already created the Jira some days ago :-)

attilapiros commented:
@agrawaldevesh  Oh you are totally right about the task end. Now we have to discuss the `lot of machinery` part.

I think it is worth to do as triggering the decommissioning is central here so having it in the main part of the test would be much more readable. You could replace this fixed time waiting with an eventually check on task ends events:
https://github.com/apache/spark/blob/913e6d72c0bb3b18739b730a0634bcbc35444437/core/src/test/scala/org/apache/spark/storage/BlockManagerDecommissionIntegrationSuite.scala#L161

Moreover if we are lucky even the `maybeDoDecommission` method with the atomic reference could be removed. The best would be to call `decommissionExecutor` from one place.

What do you think?

dagrawal3409 committed:
Incorporate @attilapiros's excellent observation

dagrawal3409 committed:
Remove some un-needed semaphores

Now that we wait for an actual task to succeed, we don't need to wait
for events prior to that: broadcast of job-info finished and task
started. The waiting for the task end/success subsumes that. Simplifying
the test even further.

SparkQA commented:
**[Test build #126657 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126657/testReport)** for PR 29226 at commit [`1d49131`](https://github.com/apache/spark/commit/1d49131fe27cf83fc0c3d62f8c3ac29024a427af).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31278/
Test PASSed.

agrawaldevesh commented:
> What do you think?

@attilapiros, as always -- an excellent observation ! Incorporating it made the code so much more leaner and everything is now in one place.

SparkQA commented:
**[Test build #126661 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126661/testReport)** for PR 29226 at commit [`bdf4d69`](https://github.com/apache/spark/commit/bdf4d693b070f3b312f296f5f27f789d6182c458).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31282/
Test PASSed.

holdenk commented on core/src/test/scala/org/apache/spark/storage/BlockManagerDecommissionIntegrationSuite.scala:
```diff
@@ -125,27 +118,26 @@ class BlockManagerDecommissionIntegrationSuite extends SparkFunSuite with LocalS
-      // Give Spark a tiny bit to start executing after the broadcast blocks land.
-      // For me this works at 100, set to 300 for system variance.
-      Thread.sleep(300)
+      // Wait for one of the tasks to succeed and finish writing its blocks.
+      // This way we know that this executor had real data to migrate when it is subsequently
+      // decommissioned below.
```
This goes against the purpose of this test: making sure that an executor with a running task that receives a decom has the block migrated. It is not migrating during decommissioning in the same way.

agrawaldevesh replied:
In which case, I don't follow how this is supposed to work in the first place. Consider the root cause: You decommission an executor while a (result) task is running. The block manager associated with the executor enters the decom state. The task tries to write the RDD persisted block. The block manager fails that because it is decom'd, thereby failing the task. The task reruns somewhere else and the job succeeds. Now you go to check if that executor has migrated the block: That check fails because the block was never written in the first place.

So I am relaxing the notion of "migrate during" to mean "migrate while the job is running". The question is "which executor" do you want to decommission and force a migration off ? Picking a wrong executor as above can mean that no migrations indeed happen because no real blocks were written (or written and thence discarded).

If that's the intent of the test then I think we need to change the block manager decommissioning (production) code to realize the intent. 

Without this PR: This test is fundamentally racy and thus provides low signal. The race is between the task's end and the block manager's decommission: If the task ends successfully before the decom, the test passes. Otherwise the test fails.

holdenk replied:
So I think another way we could make sure the test covers what we want is to run a job repeatedly until all 3 executors come up.

The block manager (in decom state) does indeed refuses puts, but RDD computation on the executor goes through `getOrElseUpdate` which immediately calls `doPutIterator` if there is not a cache hit *before* the iterator starts being computed. Since the check to see if the block manager is decommissioning occurs before the start of the computation, not at the end we want to ensure that block can be put (and then later migrated).

holdenk replied:
Maybe we should add a comment where we do the `isDecommissioning` check to explain that it is intentionally done there so that we don't reject blocks which have already started computation. Do you think that would help?

agrawaldevesh replied:
> The block manager (in decom state) does indeed refuses puts, but RDD computation on the executor goes through `getOrElseUpdate` which immediately calls `doPutIterator` if there is not a cache hit _before_ the iterator starts being computed.

I don't see this. This is what I see: (BM stands for BlockManager)
- RDD.getOrCompute -> BM.getOrElseUpdate -> BM.doPutIterator -> BM.doPut 

BM.doPut throws BlockSavedOnDecommissionedBlockManagerException if `isDecommissioning`. And this fails the job. I am not sure what it should do instead off the top of my head since I am new to this codepath. But it is certainly not continuing on with the computation (as far as I can tell).

If the intent of the test is indeed to test decommissioning "Before the task has ended", I think then we need another BlockManager PR to actually realize it :-)

holdenk replied:
I don't believe we need another block manager PR to realize it, I think this is just a flaky test because we took out the original sleep and tried to use TestUtils which doesn't do a good enough job of waiting for the executor to fully come up.

Since doPut is called *before* the task starts computation, we don't throw away any of the in-progress data.

I'll make an alternate PR to this one to illustrate my understanding and hopefully we can iron it out and make the code path clear for everyone :)

agrawaldevesh replied:
Cool ! Looking forward to your alternate PR to fix this.

> I don't believe we need another block manager PR to realize it, I think this is just a flaky test because we took out the original sleep and tried to use TestUtils which doesn't do a good enough job of waiting for the executor to fully come up.

I don't think the issue is that the executor does not come up. It does come up. The issue I think is that it tries to run the task and fails because it is decommissioned.

> Since doPut is called _before_ the task starts computation, we don't throw away any of the in-progress data.

I agree that doPut is called _before_ the task starts computation. I didn't follow what you mean by "in-progress data" ?. Per my understanding, the task is simply failing before the iterator is even created. 

Please check this for yourself by doing a decommission in the `onTaskStart` listener callback and ensuring that the task has a long enough sleep time to ensure that it waits for this decommissioning to happen.

> 
> I'll make an alternate PR to this one to illustrate my understanding and hopefully we can iron it out and make the code path clear for everyone :)

Hurray :-)

agrawaldevesh replied:
@holdenk, Please take a look at the PR once again. I have added another test to specifically capture the intent of decommissioning after the task has started but before the task has ended.

SparkQA commented:
**[Test build #126661 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126661/testReport)** for PR 29226 at commit [`bdf4d69`](https://github.com/apache/spark/commit/bdf4d693b070f3b312f296f5f27f789d6182c458).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #126657 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126657/testReport)** for PR 29226 at commit [`1d49131`](https://github.com/apache/spark/commit/1d49131fe27cf83fc0c3d62f8c3ac29024a427af).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126661/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126657/
Test PASSed.

holdenk commented:
Cool we synced up and there's two different things we believe are causing the test flakiness; neither of them should be impacting any real users but we should fix the test issues.
1) If the task starts and then the block manager is decommissioned before the task has started computing any elements the task will fail. This is OK in practice but would make this test falky.
2) Spark will, sometimes near the start of a job, not distribute the partitions equally and then we may choose to decommission an executor which had no data on it, that causes the test to fail.

This PR is going to fix #2, and I'm going to modify the BlockManager code to fix #1 in my PR.

holdenk commented:
Actually, I'm going to make a PR to this PR instead for simplifying the review process :)

holdenk commented:
https://github.com/agrawaldevesh/spark/pull/1

agrawaldevesh commented:
> * If the task starts and then the block manager is decommissioned before the task has started computing any elements the task will fail. This is OK in practice but would make this test falky.

I am toying with another idea: Instead of polling for task-start or task-end events (where task-start is defined as the time when the task starts but may not have actually begin the iterator computation), we can also just use SparkEnv.get().executorId within the task closure to get the list of tasks that have actually started computation and then explicitly decommission one of those. 

I think this approach will closely mirror the intent of being able to decommission blocks from tasks that have already started but may not have ended. This will be a test-only change without needing to change prod code.

@holdenk, let me know what you think of this approach.

holdenk commented:
Sounds good to me 👍

dagrawal3409 committed:
Added 2 task start tests: One of which we expect to fail and one that captures @holdenk's intent

SparkQA commented:
**[Test build #126678 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126678/testReport)** for PR 29226 at commit [`1864347`](https://github.com/apache/spark/commit/1864347648f64ae4bf685feb358f52540e1a1561).

agrawaldevesh commented on core/src/test/scala/org/apache/spark/storage/BlockManagerDecommissionIntegrationSuite.scala:
```diff
@@ -36,21 +37,34 @@ class BlockManagerDecommissionIntegrationSuite extends SparkFunSuite with LocalS
   val numExecs = 3
   val numParts = 3
 
+  test(s"verify that an already running task which is going to cache data succeeds " +
+    s"on a decommissioned executor after task start") {
+    runDecomTest(true, false, true, 1)
+  }
+
+  test(s"verify that an already running task which is going to cache data succeeds " +
+    s"on a decommissioned executor after iterator start") {
+    runDecomTest(true, false, true, 2)
+  }
+
```
@holdenk .. this is what I meant: I have added two tests to test the decom after "Task Start". 

The first test is a negative test. It defines task start as when the driver says that the task is "launched" and as we discussed, it is expected to fail because decommissioning will be triggered before `BlockManager.doPut` is called and thus fail the task without starting the iterator.

The second test is what you want mentioned in your "intention": We want the decommissioning after the "iterator start" (ie user code started) to trigger a migration: The block manager will write the block anyway and it will be migrated.

The first negative test is just to demonstrate that it is a case we don't handle yet. Since we agree that it is not a case worth fixing for production, I will eventually delete it. 

So now we cover the full gamut of interesting decom points in task: When it is just launched, When user code is activated, After it has ended.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31299/
Test PASSed.

dagrawal3409 committed:
Clean it up: Remove the negative test which we know will fail

SparkQA commented:
**[Test build #126681 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126681/testReport)** for PR 29226 at commit [`b98495c`](https://github.com/apache/spark/commit/b98495cdeb0bf64a1e7997c9d9867e617ce058f3).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31302/
Test PASSed.

dagrawal3409 committed:
Tweak sleep times depending on when to decom

SparkQA commented:
**[Test build #126684 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126684/testReport)** for PR 29226 at commit [`070d30c`](https://github.com/apache/spark/commit/070d30ce84735a3483901616fb972b69f1b096ce).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31305/
Test PASSed.

dagrawal3409 committed:
closure cleaner fix

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31308/
Test PASSed.

SparkQA commented:
**[Test build #126687 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126687/testReport)** for PR 29226 at commit [`bea7776`](https://github.com/apache/spark/commit/bea777668f1494860a1c7588e0567028b1611c00).

SparkQA commented:
**[Test build #126678 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126678/testReport)** for PR 29226 at commit [`1864347`](https://github.com/apache/spark/commit/1864347648f64ae4bf685feb358f52540e1a1561).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126678/
Test FAILed.

SparkQA commented:
**[Test build #126684 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126684/testReport)** for PR 29226 at commit [`070d30c`](https://github.com/apache/spark/commit/070d30ce84735a3483901616fb972b69f1b096ce).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126684/
Test FAILed.

SparkQA commented:
**[Test build #126687 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126687/testReport)** for PR 29226 at commit [`bea7776`](https://github.com/apache/spark/commit/bea777668f1494860a1c7588e0567028b1611c00).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #126681 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126681/testReport)** for PR 29226 at commit [`b98495c`](https://github.com/apache/spark/commit/b98495cdeb0bf64a1e7997c9d9867e617ce058f3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126687/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126681/
Test PASSed.

agrawaldevesh commented:
> [agrawaldevesh#1](https://github.com/agrawaldevesh/spark/pull/1)

I think this isn't needed anymore.

holdenk commented on core/src/test/scala/org/apache/spark/storage/BlockManagerDecommissionIntegrationSuite.scala:
```diff
@@ -87,36 +113,30 @@ class BlockManagerDecommissionIntegrationSuite extends SparkFunSuite with LocalS
-    sc.addSparkListener(new SparkListener {
 
+    def getCandidateExecutorToDecom: Option[String] = if (whenToDecom == TaskStarted) {
+      accum.value.asScala.headOption
```
I don't think this is going to work as intended, accumulators send updates back at the end of the task, unless something has changed.

holdenk replied:
If you don't believe that is the case, can we add an assertion in here that none of the tasks have finished yet?

holdenk commented:
I think I understand your intention with accumulators now, I don't think that's going to work.

dagrawal3409 committed:
Accumulator didn't quite work out of the box but sort of works

SparkQA commented:
**[Test build #126743 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126743/testReport)** for PR 29226 at commit [`ed8224b`](https://github.com/apache/spark/commit/ed8224be8f539f71e0233da16ae9d03819e584be).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31366/
Test PASSed.

agrawaldevesh commented:
> I think I understand your intention with accumulators now, I don't think that's going to work.

Thanks for pointing this out @holdenk. As a spark n00b I have certainly learnt quite a bit by fixing this test. I tried another approach and also added the asserts you asked for. 

I learnt that while spark does not update the accumulator until the task completes, it does graciously trigger a listener event when the accumulators are received from an executor via the heartbeat mechanism. I am using that to know for sure that the task has started for "real".

Please take a look.

SparkQA commented:
**[Test build #126744 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126744/testReport)** for PR 29226 at commit [`1b469fb`](https://github.com/apache/spark/commit/1b469fb0101730916c9327d22cf5b3e27bf4eadb).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31368/
Test PASSed.

SparkQA commented:
**[Test build #126743 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126743/testReport)** for PR 29226 at commit [`ed8224b`](https://github.com/apache/spark/commit/ed8224be8f539f71e0233da16ae9d03819e584be).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126743/
Test PASSed.

SparkQA commented:
**[Test build #126744 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126744/testReport)** for PR 29226 at commit [`1b469fb`](https://github.com/apache/spark/commit/1b469fb0101730916c9327d22cf5b3e27bf4eadb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

holdenk commented:
Yeah if you wait long enough for a heartbeat and check the event that should do the trick, but to avoid a flaky test your going to need a good length of time. I'll take another look tomorrow afternoon (have meetings all morning).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126744/
Test PASSed.

holdenk reviewed (approved):
LGTM. If no other comments I'll merge this tomorrow.

HyukjinKwon commented:
For some reasons, the test seems still flaky: https://github.com/apache/spark/pull/29373/checks?check_run_id=953464967

```
BlockManagerDecommissionIntegrationSuite:
- verify that an already running task which is going to cache data succeeds on a decommissioned executor after task start *** FAILED *** (13 seconds, 145 milliseconds)
  The code passed to eventually never returned normally. Attempted 1829 times over 6.001362179 seconds. Last failure message: getCandidateExecutorToDecom.isDefined was false. (BlockManagerDecommissionIntegrationSuite.scala:179)
  org.scalatest.exceptions.TestFailedDueToTimeoutException:
  at org.scalatest.enablers.Retrying$$anon$4.tryTryAgain$2(Retrying.scala:185)
  at org.scalatest.enablers.Retrying$$anon$4.retry(Retrying.scala:192)
  at org.scalatest.concurrent.Eventually.eventually(Eventually.scala:402)
  at org.scalatest.concurrent.Eventually.eventually$(Eventually.scala:401)
  at org.apache.spark.storage.BlockManagerDecommissionIntegrationSuite.eventually(BlockManagerDecommissionIntegrationSuite.scala:34)
  at org.scalatest.concurrent.Eventually.eventually(Eventually.scala:312)
  at org.scalatest.concurrent.Eventually.eventually$(Eventually.scala:311)
  at org.apache.spark.storage.BlockManagerDecommissionIntegrationSuite.eventually(BlockManagerDecommissionIntegrationSuite.scala:34)
  at org.apache.spark.storage.BlockManagerDecommissionIntegrationSuite.runDecomTest(BlockManagerDecommissionIntegrationSuite.scala:179)
  at org.apache.spark.storage.BlockManagerDecommissionIntegrationSuite.$anonfun$new$1(BlockManagerDecommissionIntegrationSuite.scala:45)
  at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
  at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
  at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
  at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
  at org.scalatest.Transformer.apply(Transformer.scala:22)
  at org.scalatest.Transformer.apply(Transformer.scala:20)
  at org.scalatest.funsuite.AnyFunSuiteLike$$anon$1.apply(AnyFunSuiteLike.scala:189)
  at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:158)
  at org.scalatest.funsuite.AnyFunSuiteLike.invokeWithFixture$1(AnyFunSuiteLike.scala:187)
  at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTest$1(AnyFunSuiteLike.scala:199)
  at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
  at org.scalatest.funsuite.AnyFunSuiteLike.runTest(AnyFunSuiteLike.scala:199)
  at org.scalatest.funsuite.AnyFunSuiteLike.runTest$(AnyFunSuiteLike.scala:181)
  at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterEach$$super$runTest(SparkFunSuite.scala:60)
  at org.scalatest.BeforeAndAfterEach.runTest(BeforeAndAfterEach.scala:234)
  at org.scalatest.BeforeAndAfterEach.runTest$(BeforeAndAfterEach.scala:227)
  at org.apache.spark.SparkFunSuite.runTest(SparkFunSuite.scala:60)
  at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTests$1(AnyFunSuiteLike.scala:232)
  at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:413)
  at scala.collection.immutable.List.foreach(List.scala:392)
  at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
  at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:396)
  at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:475)
  at org.scalatest.funsuite.AnyFunSuiteLike.runTests(AnyFunSuiteLike.scala:232)
  at org.scalatest.funsuite.AnyFunSuiteLike.runTests$(AnyFunSuiteLike.scala:231)
  at org.scalatest.funsuite.AnyFunSuite.runTests(AnyFunSuite.scala:1562)
  at org.scalatest.Suite.run(Suite.scala:1112)
  at org.scalatest.Suite.run$(Suite.scala:1094)
  at org.scalatest.funsuite.AnyFunSuite.org$scalatest$funsuite$AnyFunSuiteLike$$super$run(AnyFunSuite.scala:1562)
  at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$run$1(AnyFunSuiteLike.scala:236)
  at org.scalatest.SuperEngine.runImpl(Engine.scala:535)
  at org.scalatest.funsuite.AnyFunSuiteLike.run(AnyFunSuiteLike.scala:236)
  at org.scalatest.funsuite.AnyFunSuiteLike.run$(AnyFunSuiteLike.scala:235)
  at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterAll$$super$run(SparkFunSuite.scala:60)
  at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
  at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
  at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
  at org.apache.spark.SparkFunSuite.run(SparkFunSuite.scala:60)
  at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:318)
  at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:513)
  at sbt.ForkMain$Run$2.call(ForkMain.java:296)
  at sbt.ForkMain$Run$2.call(ForkMain.java:286)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
  at java.lang.Thread.run(Thread.java:748)
  Cause: org.scalatest.exceptions.TestFailedException: getCandidateExecutorToDecom.isDefined was false
  at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:472)
  at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:471)
  at org.scalatest.Assertions$.newAssertionFailedException(Assertions.scala:1231)
  at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:1295)
  at org.apache.spark.storage.BlockManagerDecommissionIntegrationSuite.$anonfun$runDecomTest$6(BlockManagerDecommissionIntegrationSuite.scala:180)
  at org.scalatest.enablers.Retrying$$anon$4.makeAValiantAttempt$1(Retrying.scala:150)
  at org.scalatest.enablers.Retrying$$anon$4.tryTryAgain$2(Retrying.scala:162)
  at org.scalatest.enablers.Retrying$$anon$4.retry(Retrying.scala:192)
  at org.scalatest.concurrent.Eventually.eventually(Eventually.scala:402)
  at org.scalatest.concurrent.Eventually.eventually$(Eventually.scala:401)
  at org.apache.spark.storage.BlockManagerDecommissionIntegrationSuite.eventually(BlockManagerDecommissionIntegrationSuite.scala:34)
  at org.scalatest.concurrent.Eventually.eventually(Eventually.scala:312)
  at org.scalatest.concurrent.Eventually.eventually$(Eventually.scala:311)
  at org.apache.spark.storage.BlockManagerDecommissionIntegrationSuite.eventually(BlockManagerDecommissionIntegrationSuite.scala:34)
  at org.apache.spark.storage.BlockManagerDecommissionIntegrationSuite.runDecomTest(BlockManagerDecommissionIntegrationSuite.scala:179)
  at org.apache.spark.storage.BlockManagerDecommissionIntegrationSuite.$anonfun$new$1(BlockManagerDecommissionIntegrationSuite.scala:45)
  at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
  at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
  at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
  at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
  at org.scalatest.Transformer.apply(Transformer.scala:22)
  at org.scalatest.Transformer.apply(Transformer.scala:20)
  at org.scalatest.funsuite.AnyFunSuiteLike$$anon$1.apply(AnyFunSuiteLike.scala:189)
  at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:158)
  at org.scalatest.funsuite.AnyFunSuiteLike.invokeWithFixture$1(AnyFunSuiteLike.scala:187)
  at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTest$1(AnyFunSuiteLike.scala:199)
  at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
  at org.scalatest.funsuite.AnyFunSuiteLike.runTest(AnyFunSuiteLike.scala:199)
  at org.scalatest.funsuite.AnyFunSuiteLike.runTest$(AnyFunSuiteLike.scala:181)
  at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterEach$$super$runTest(SparkFunSuite.scala:60)
  at org.scalatest.BeforeAndAfterEach.runTest(BeforeAndAfterEach.scala:234)
  at org.scalatest.BeforeAndAfterEach.runTest$(BeforeAndAfterEach.scala:227)
  at org.apache.spark.SparkFunSuite.runTest(SparkFunSuite.scala:60)
  at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTests$1(AnyFunSuiteLike.scala:232)
  at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:413)
  at scala.collection.immutable.List.foreach(List.scala:392)
  at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
  at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:396)
  at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:475)
  at org.scalatest.funsuite.AnyFunSuiteLike.runTests(AnyFunSuiteLike.scala:232)
  at org.scalatest.funsuite.AnyFunSuiteLike.runTests$(AnyFunSuiteLike.scala:231)
  at org.scalatest.funsuite.AnyFunSuite.runTests(AnyFunSuite.scala:1562)
  at org.scalatest.Suite.run(Suite.scala:1112)
  at org.scalatest.Suite.run$(Suite.scala:1094)
  at org.scalatest.funsuite.AnyFunSuite.org$scalatest$funsuite$AnyFunSuiteLike$$super$run(AnyFunSuite.scala:1562)
  at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$run$1(AnyFunSuiteLike.scala:236)
  at org.scalatest.SuperEngine.runImpl(Engine.scala:535)
  at org.scalatest.funsuite.AnyFunSuiteLike.run(AnyFunSuiteLike.scala:236)
  at org.scalatest.funsuite.AnyFunSuiteLike.run$(AnyFunSuiteLike.scala:235)
  at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterAll$$super$run(SparkFunSuite.scala:60)
  at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
  at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
  at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
  at org.apache.spark.SparkFunSuite.run(SparkFunSuite.scala:60)
  at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:318)
  at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:513)
  at sbt.ForkMain$Run$2.call(ForkMain.java:296)
  at sbt.ForkMain$Run$2.call(ForkMain.java:286)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
  at java.lang.Thread.run(Thread.java:748)
```
